### PR TITLE
HIP CPU test_ops CI

### PR DESCRIPTION
**Pull request by tralamazza**
HIP CPU test_ops CI based on @dc-dc-dc #1419.

`HIP=1 HIP_CPU=1 python -m pytest -v test/test_ops.p` takes ~11 min on a thinkpad T480s, w/ an average build time ~200ms per kernel.

missing:
- [x] fix clang racing (probably needs a file lock or file per pid)
- [x] removing one trampoline (just need to pass another argument)
- [x] make the linter happy?
- [x] check w/ flamegraph the average kernel build time

bounty ref: #1700 

shout-out to [corecode](https://github.com/corecode/)

### CODE REVIEW

File: extra/hip_wrapper.py [hunk 1]
```diff
@@ -681,5 +680,9 @@ def hiprtcGetCode(prog):
     status = _libhiprtc.hiprtcGetCode(prog, e_code)
     hipCheckStatus(status)
     return e_code
+
+  def language(**kwargs):
```
Comment (geohot) [inline extra/hip_wrapper.py:684]: This shouldn't be in here, it's just hiding complexity.

Comment (tralamazza) [reply extra/hip_wrapper.py:684]: I tried to minimize the impact on ops_hips.py, do you have any suggestions?

Comment (geohot) [reply extra/hip_wrapper.py:684]: Just put it in ops_hip. If it's required complexity that's okay, but it shouldn't hide

Comment (tralamazza) [reply extra/hip_wrapper.py:684]: but then ops_hip will need another "if HIP_CPU"

Comment (tralamazza) [reply extra/hip_wrapper.py:684]: unfortunately HIP-CPU is a bit of a mess and lacks a few APIs (e.g. capture), maybe HIP_CPU should become a Device? so tests can filtered it out.

Comment (geohot) [reply extra/hip_wrapper.py:684]: If it's a different Device there's no point in doing this, what's being tested?

Comment (geohot) [reply extra/hip_wrapper.py:684]: "but then ops_hip will need another "if HIP_CPU""

and...? if you hide complexity, you still have the complexity, it's just elsewhere. surfaced complexity > hidden complexity.

Comment (tralamazza) [reply extra/hip_wrapper.py:684]: > "but then ops_hip will need another "if HIP_CPU""
> 
> and...? if you hide complexity, you still have the complexity, it's just elsewhere. surfaced complexity > hidden complexity.

It's an interface provided by a hip cpu wrapper, like all those hip_ functions, implemented by both wrappers. Should I move all implementations to ops_hip?

Comment (tralamazza) [reply extra/hip_wrapper.py:684]: > If it's a different Device there's no point in doing this, what's being tested?

test_ops is being tested

File: tinygrad/runtime/ops_hip.py [hunk 2]
```diff
@@ -100,6 +104,7 @@ def __init__(self, name:str, prg:str, binary=False):
       self.modules.append(hip.hipModuleLoadData(prg))
       self.prgs.append(hip.hipModuleGetFunction(self.modules[-1], name))
 
+  cache_compiled = lambda _: _ if getenv("HIP_CPU") else cache_compiled # noqa: E731
```
Comment (geohot) [inline tinygrad/runtime/ops_hip.py:107]: Why does the cache break with CPU?

File: tinygrad/runtime/ops_hip.py [hunk 3]
```diff
@@ -127,7 +132,8 @@ class PackageStruct(ctypes.Structure):
   def __del__(self):
     for module in self.modules: hip.hipModuleUnload(module)
 
-renderer = functools.partial(uops_to_cstyle, CStyleLanguage(
+language = hip.language if getenv("HIP_CPU") else CStyleLanguage
```
Comment (geohot) [inline tinygrad/runtime/ops_hip.py:135]: I'm fine with this, at least the normal HIP is here

**Comment by geohot** [comment]
Looking good code wise, now we just have to do something about the speed. How is this slower than GPUocelot?!

File: tinygrad/runtime/ops_hip.py (hunk 2 above)
Comment (tralamazza) [reply tinygrad/runtime/ops_hip.py:107]: because hiprtcCreateProgram returns the CDLL (and write_bytes expects byte-like objects)

**Comment by tralamazza** [comment]
Locally `HIP=1 HIP_CPU=1 python -m pytest -v -n=auto test/test_ops.py` takes:

- 1st pass (full build and run)
`199 passed, 8 skipped in 224.93s (0:03:44)`

- 2nd pass (no build needed since /tmp contains all binaries)
`199 passed, 8 skipped in 62.75s (0:01:02)`

**Comment by tralamazza** [comment]
> Looking good code wise, now we just have to do something about the speed. How is this slower than GPUocelot?!

I haven't profiled so I don't know yet.

Executing locally with all kernels precompiled still takes ~1min, pretty slow for just loading a shared lib and executing a function.

**Comment by geohot** [comment]
Closing, CI is too slow to merge. You are welcome to reopen when it's ~5 minutes!

**Pull request closed without merging**